core/modules/loader/ServerUdpBase.cc

```diff
@@ -0,0 +1,150 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2018 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+// Class header
+#include "loader/ServerUdpBase.h"
+
+// System headers
+#include <iostream>
+#include <unistd.h>
+
+// Third-party headers
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.loader.ServerUdpBase");
+}
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+std::atomic<uint64_t> ServerUdpBase::_msgIdSeq{1};
+
+
+ServerUdpBase::ServerUdpBase(boost::asio::io_service& io_service, std::string const& host, int port)
+    : _ioService(io_service),
+      _socket(io_service, boost::asio::ip::udp::endpoint(boost::asio::ip::udp::v4(), port)),
+      _hostName(host), _port(port) {
+    _receivePrepare(); // Prime the server for an incoming message.
+}
+
+
+void ServerUdpBase::_receiveCallback(boost::system::error_code const& error, size_t bytesRecvd) {
+    _data->advanceWriteCursor(bytesRecvd); // _data needs to know the valid portion of the buffer.
+    if (!error && bytesRecvd > 0) {
+        LOGS(_log, LOG_LVL_INFO, "rCb received(" << bytesRecvd << "):" <<
+                                 ", code=" << error << ", from endpoint=" << _senderEndpoint);
+
+        _sendData = parseMsg(_data, _senderEndpoint);
+        if (_sendData != nullptr) {
+            _sendResponse();
+        } else {
+            _receivePrepare();
+        }
+    } else {
+        LOGS(_log, LOG_LVL_ERROR, "ServerUdpBase::_receiveCallback got empty message, ignoring");
```

Comment:
That's not technically true, because this also cover a case of a error in the communication. Hence the right logic should be:
```
if (error) {
  # extract error message for the error and report it.
} else if (bytesRecvd == 0) {
  # complain about 0-length message like you do
} else {
  # normal processing
}
```
Please, revisit the rest of your code to see if you have the same problem in there as well.